server/controllers/finance/reports/income_expense_by_year/index.js

```diff
@@ -0,0 +1,308 @@
+/**
+ * Income Expense by Month Controller
+ *
+ * This controller is responsible for processing the income and expense by month
+ * report.
+ *
+ * @requires lodash
+ * @requires lib/db
+ * @requires lib/ReportManager
+ * @requires lib/errors/BadRequest
+ * @requires lib/Tree
+ * @requires controllers/fiscal
+ */
+
+
+const _ = require('lodash');
+const db = require('../../../../lib/db');
+const ReportManager = require('../../../../lib/ReportManager');
+const Tree = require('../../../../lib/Tree');
+const util = require('../../../../lib/util');
+
+const TEMPLATE = './server/controllers/finance/reports/income_expense_by_year/report.handlebars';
+
+const INCOME_TYPE_ID = 4;
+const EXPENSE_TYPE_ID = 5;
+const TITLE_ID = 6;
+
+const DEFAULT_PARAMS = {
+  csvKey : 'rows',
+  filename : 'TREE.INCOME_EXPENSE_BY_MONTH',
+  orientation : 'landscape',
+  footerRight : '[page] / [toPage]',
+};
+
+// expose to the API
+exports.document = document;
+
+async function document(req, res, next) {
+  let report;
+
+  const options = _.defaults(req.query, DEFAULT_PARAMS);
+
+  const removeUnusedAccounts = !!+options.removeUnusedAccounts;
+
+  try {
+    report = new ReportManager(TEMPLATE, req.session, options);
+  } catch (e) {
+    next(e);
+    return;
+  }
+
+  const { fiscalYearId } = options;
+  const data = {};
+
+
+  const thirdYear = await getFiscalById(fiscalYearId);
+  const secondYear = await getFiscalById(thirdYear.previous_fiscal_year_id);
+  const firstYear = await getFiscalById(secondYear.previous_fiscal_year_id);
+
+  if (!firstYear || !secondYear || !thirdYear) {
+    res.status(500).json({ msg : `bad fiscal year's range` });
+    return;
+  }
+  _.extend(data, {
+    firstYear,
+    secondYear,
+    thirdYear,
+  });
+
+
+  Promise.all([
```

Comment:
Since you already started with `async/await`, why not make it all `async/await`? 

```js
try {

  const [firstBalances, secondBalances, thirdBalances] = await Promise.all([
    getAccountBalances(firstYear.id),
     // etc
   ]);
 
   const dataset =  combineIntoSingleDataset(secondBalances, firstBalances, thirdBalances);
   // etc  ...
} catch (e) {
  next(e);
}
```